nav2_bt_navigator/src/bt_navigator.cpp

```diff
@@ -231,10 +231,27 @@ BtNavigator::onLoop()
 }
 
 void
-BtNavigator::onPreempt()
+BtNavigator::onPreempt(Action::Goal::ConstSharedPtr goal)
 {
   RCLCPP_INFO(get_logger(), "Received goal preemption request");
-  initializeGoalPose(bt_action_server_->acceptPendingGoal());
+
+  if (goal->behavior_tree == bt_action_server_->getCurrentBTFilename() ||
+    (goal->behavior_tree.empty() &&
+    bt_action_server_->getCurrentBTFilename() == bt_action_server_->getDefaultBTFilename()))
+  {
+    // if pending goal requests the same BT as the current goal, accept the pending goal
+    // if pending goal has an empty behavior_tree field, it requests the default BT file
+    // accept the pending goal if the current goal is running the default BT file
+    initializeGoalPose(bt_action_server_->acceptPendingGoal());
+  } else {
+    RCLCPP_WARN(
+      get_logger(),
+      "Preemption request was rejected since the requested BT XML file is not the same "
+      "as the one that the current goal is executing. Cancel the current goal and send a new "
```

Comment:
A short sentence stating "why" would be nice, something like: Preemption with a new BT is invalid since it would require pausing task execution, not "preempting".